symengine/as_real_imag.cpp

```diff
@@ -0,0 +1,284 @@
+        }
+    }
+
+    void bvisit(const Add &x)
+    {
+        RCP<const Basic> re_, im_, t;
+        umap_basic_num dr, dim;
+        RCP<const Number> coefr = zero, coefim = zero, coef;
+
+        for (const auto &arg : x.get_args()) {
+            as_real_imag(arg, outArg(re_), outArg(im_));
+            if (is_a_Number(*re_)) {
+                iaddnum(outArg(coefr), rcp_static_cast<const Number>(re_));
+            } else {
+                Add::as_coef_term(re_, outArg(coef), outArg(t));
+                Add::dict_add_term(dr, coef, t);
+            }
+            if (is_a_Number(*im_)) {
+                iaddnum(outArg(coefim), rcp_static_cast<const Number>(im_));
+            } else {
+                Add::as_coef_term(im_, outArg(coef), outArg(t));
+                Add::dict_add_term(dim, coef, t);
+            }
+        }
+
+        *real_ = Add::from_dict(coefr, std::move(dr));
+        *imag_ = Add::from_dict(coefim, std::move(dim));
+    }
+
+    void bvisit(const Pow &x)
+    {
+        RCP<const Basic> base_, exp_, re, im;
+        base_ = x.get_base();
+        exp_ = x.get_exp();
+        as_real_imag(base_, outArg(re), outArg(im));
+
+        if (eq(*im, *zero)) {
+            *real_ = x.rcp_from_this();
+            *imag_ = zero;
+            return;
+        }
+        if (is_a<Integer>(*exp_)) {
+            RCP<const Basic> expx;
+            if (eq(*Gt(exp_, zero), *boolTrue)) {
+                expx = expand(x.rcp_from_this());
+            } else {
+                auto magn = add(mul(re, re), mul(im, im));
+                re = div(re, magn);
+                im = div(neg(im), magn);
+                expx = expand(pow(add(re, mul(im, I)), neg(exp_)));
+            }
+            if (eq(*expx, x))
+                throw SymEngineException("Not Implemented");
+            as_real_imag(expx, real_, imag_);
+        } else if (is_a<Rational>(*exp_)) {
+            auto magn = sqrt(add(mul(re, re), mul(im, im)));
+            auto ang = atan2(im, re);
+            magn = pow(magn, exp_);
+            ang = mul(ang, exp_);
+            *real_ = mul(magn, cos(ang));
+            *imag_ = mul(magn, sin(ang));
+        } else {
+            throw SymEngineException("Not Implemented");
+        }
+    }
+
+    void bvisit(const ComplexBase &x)
+    {
+        *real_ = x.real_part();
+        *imag_ = x.imaginary_part();
+    }
+
+    void bvisit(const Infty &x)
+    {
+        if (eq(x, *ComplexInf)) {
+            *real_ = Nan;
+            *imag_ = Nan;
+        } else {
+            *real_ = x.rcp_from_this();
+            *imag_ = zero;
+        }
+    }
+
+    void bvisit(const Sin &x)
+    {
+        RCP<const Basic> re_, im_;
+        as_real_imag(x.get_arg(), outArg(re_), outArg(im_));
+        *real_ = mul(sin(re_), cosh(im_));
+        *imag_ = mul(sinh(im_), cos(re_));
+    }
+
+    void bvisit(const Cos &x)
+    {
+        RCP<const Basic> re_, im_;
+        as_real_imag(x.get_arg(), outArg(re_), outArg(im_));
```

Comment:
Calling `as_real_imag` here makes it impossible to extend this visitor. You should call `apply(*x.get_arg())` here and then `real_` and `imag_` would be set to the arguments you need.